File: backend/tests/config_test.py
"""Unit tests for `Config` and `load_config`."""
from __future__ import annotations

import pytest

from dtm_backend.config import (
    DEFAULT_CORS_ORIGINS,
    DEFAULT_RPC_FALLBACKS,
    Config,
    load_config,
)


def test_config_defaults() -> None:
    """Sensible defaults for an empty environment."""
    cfg = Config()
    assert cfg.port == 8000
    assert cfg.host == "0.0.0.0"
    assert cfg.log_level == "info"
    assert cfg.rpc_url == DEFAULT_RPC_FALLBACKS[0]
    assert cfg.rpc_ws_url is None
    assert cfg.chain_id == 1
    assert cfg.cache_ttl_ms == 5_000
    assert cfg.liquidation_poll_ms == 12_000
    assert cfg.liquidation_lookback == 100
    assert cfg.amm_sync_poll_ms == 12_000
    assert cfg.amm_sync_lookback == 100
    assert cfg.amm_sync_debounce_ms == 250
    assert cfg.cors_origins == DEFAULT_CORS_ORIGINS
    assert cfg.cors_allow_all is False
    assert cfg.env == "dev"


def test_config_cors_origin_list_returns_list() -> None:
    """`cors_origin_list()` is always a fresh list (not a tuple)."""
    cfg = Config()
    assert isinstance(cfg.cors_origin_list(), list)
    assert cfg.cors_origin_list() == list(DEFAULT_CORS_ORIGINS)


def test_config_env_override(monkeypatch: pytest.MonkeyPatch) -> None:
    """`PORT=9000` in the env propagates to `Config.port`."""
    monkeypatch.setenv("PORT", "9000")
    monkeypatch.setenv("CHAIN_ID", "11155111")
    monkeypatch.setenv("CORS_ALLOW_ALL", "true")
    cfg = load_config()
    assert cfg.port == 9000
    assert cfg.chain_id == 11155111
    assert cfg.cors_allow_all is True


def test_config_extra_env_keys_are_ignored(monkeypatch: pytest.MonkeyPatch) -> None:
    """Unknown env keys must not raise (extra='ignore')."""
    monkeypatch.setenv("SOMETHING_COMPLETELY_UNKNOWN", "ok")
    cfg = load_config()
    assert cfg.port == 8000  # default preserved

File: backend/tests/config.test.ts
/**
 * Config tests — focused on the CORS allow-list resolution and the
 * `CORS_ORIGINS` env var parsing.
 *
 * The rest of the env-var readers are exercised indirectly by the
 * `loadConfig` integration test (it would require a heavy mock of
 * `process.env` for marginal value), so this file sticks to the
 * behaviour that is new in this PR.
 */
import { describe, expect, it } from 'vitest';

import { loadConfig, __TEST_DEFAULT_CORS_ORIGINS } from '../src/config.js';

const ENV_BASE: NodeJS.ProcessEnv = {
  // Set the other readers to deterministic values; we only care
  // about the CORS plumbing in this file.
  RPC_URL: 'http://example.invalid',
  CHAIN_ID: '1',
  LOG_LEVEL: 'silent',
  PORT: '8000',
  CACHE_TTL_MS: '1000',
  LIQUIDATION_POLL_MS: '1000',
  LIQUIDATION_LOOKBACK: '10',
  AMM_SYNC_POLL_MS: '1000',
  AMM_SYNC_LOOKBACK: '10',
  AMM_SYNC_DEBOUNCE_MS: '50',
};

describe('loadConfig — CORS', () => {
  it('defaults to the dev loopback list when CORS_ORIGINS is unset', () => {
    const cfg = loadConfig({ ...ENV_BASE });
    expect(cfg.corsAllowAll).toBe(false);
    expect(cfg.corsOrigins).toEqual(__TEST_DEFAULT_CORS_ORIGINS);
    expect(cfg.corsOrigins).toContain('http://localhost:5173');
    expect(cfg.corsOrigins).toContain('http://127.0.0.1:5173');
  });

  it('treats an empty CORS_ORIGINS as the dev defaults', () => {
    const cfg = loadConfig({ ...ENV_BASE, CORS_ORIGINS: '' });
    expect(cfg.corsAllowAll).toBe(false);
    expect(cfg.corsOrigins).toEqual(__TEST_DEFAULT_CORS_ORIGINS);
  });

  it('treats the literal "*" as allow-all', () => {
    const cfg = loadConfig({ ...ENV_BASE, CORS_ORIGINS: '*' });
    expect(cfg.corsAllowAll).toBe(true);
    expect(cfg.corsOrigins).toEqual(['*']);
  });

  it('parses a comma-separated allow-list verbatim (trimmed, deduped)', () => {
    const cfg = loadConfig({
      ...ENV_BASE,
      CORS_ORIGINS: 'https://a.example, https://b.example,https://a.example,  ',
    });
    expect(cfg.corsAllowAll).toBe(false);
    expect(cfg.corsOrigins).toEqual(['https://a.example', 'https://b.example']);
  });

  it('drops entries that are empty after trimming', () => {
    const cfg = loadConfig({
      ...ENV_BASE,
      CORS_ORIGINS: ',,,https://a.example,,',
    });
    expect(cfg.corsOrigins).toEqual(['https://a.example']);
  });
});
